src/transformers/models/canine/modeling_canine.py

```diff
@@ -0,0 +1,1559 @@
+
+        # We also keep [CLS] as a separate sequence position since we always
+        # want to reserve a position (and the model capacity that goes along
+        # with that) in the deep BERT stack.
+        # `result`: [batch, molecule_seq, molecule_dim]
+        result = torch.cat([cls_encoding, downsampled_truncated], dim=1)
+
+        result = self.LayerNorm(result)
+
+        return result
+
+
+class ConvProjection(nn.Module):
+    """
+    Project representations from hidden_size*2 back to hidden_size across a window of w = config.upsampling_kernel_size
+    characters.
+    """
+
+    def __init__(self, config):
+        super().__init__()
+        self.config = config
+        self.conv = nn.Conv1d(
+            in_channels=config.hidden_size * 2,
+            out_channels=config.hidden_size,
+            kernel_size=config.upsampling_kernel_size,
+            stride=1,
+        )
+        self.activation = ACT2FN[config.hidden_act]
+        # self.LayerNorm is not snake-cased to stick with TensorFlow model variable name and be able to load
+        # any TensorFlow checkpoint file
+        self.LayerNorm = nn.LayerNorm(config.hidden_size, eps=config.layer_norm_eps)
+        self.dropout = nn.Dropout(config.hidden_dropout_prob)
+
+    def forward(self, inputs, final_seq_char_positions=None):
+        # inputs has shape [batch, mol_seq, molecule_hidden_size+char_hidden_final]
+        # we transpose it to be [batch, molecule_hidden_size+char_hidden_final, mol_seq]
+        inputs = torch.transpose(inputs, 1, 2)
+
+        # PyTorch < 1.9 does not support padding="same" (which is used in the original implementation),
+        # so we pad the tensor manually before passing it to the conv layer
+        # based on https://github.com/google-research/big_transfer/blob/49afe42338b62af9fbe18f0258197a33ee578a6b/bit_tf2/models.py#L36-L38
+        pad_total = self.config.upsampling_kernel_size - 1
+        pad_beg = pad_total // 2
+        pad_end = pad_total - pad_beg
+
+        pad = nn.ConstantPad1d((pad_beg, pad_end), 0)
+        # `result`: shape (batch_size, char_seq_len, hidden_size)
+        result = self.conv(pad(inputs))
+        result = torch.transpose(result, 1, 2)
+        result = self.activation(result)
+        result = self.LayerNorm(result)
+        result = self.dropout(result)
+        final_char_seq = result
+
+        if final_seq_char_positions is not None:
+            # Limit transformer query seq and attention mask to these character
+            # positions to greatly reduce the compute cost. Typically, this is just
+            # done for the MLM training task.
+
+            # `query_seq`: [batch, final_char_seq, char_dim]
+            # query_seq = tf.gather(
+            #     final_char_seq, final_seq_char_positions, batch_dims=1)
+            # char_attention_mask = tf.gather(
+            #     char_attention_mask,
+            #     final_seq_char_positions,
+            #     batch_dims=1)
```

Comment:
Should this be cleaned? Or is it needed with the TODO?

Reply:
It's a todo, I raise a `NotImplementedError` for now.